controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java

```diff
@@ -102,12 +143,31 @@ protected synchronized boolean tryAcquire(int permit) {
         }
 
         @Override
-        protected synchronized boolean tryRelease(int permit) {
-            final Object value = permitHolder.get();
-            if (value == null) {
+        protected int tryAcquireShared(final int permit) {
+            if (permitHolder != null) {
+                return -1;
+            }
+            if (compareAndSetState(0, 1)) {
+                return 1;
+            } else {
```

Comment:
Sorry I missed this with all the other spam from that PR. I realized this
after I'd already found AQLS, then re-read Jason's comment after I'd done
it ;)

I ended up closing this PR, new one is:
https://github.com/wildfly/wildfly-core/pull/1457
On Mar 11, 2016 4:43 PM, "Brian Stansberry" <notifications@github.com>
wrote:

> In
> controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java
> <https://github.com/wildfly/wildfly-core/pull/1453#discussion_r55900166>:
>
> > @@ -102,12 +143,31 @@ protected synchronized boolean tryAcquire(int permit) {
> >          }
> >
> >          @Override
> > -        protected synchronized boolean tryRelease(int permit) {
> > -            final Object value = permitHolder.get();
> > -            if (value == null) {
> > +        protected int tryAcquireShared(final int permit) {
> > +            if (permitHolder != null) {
> > +                return -1;
> > +            }
> > +            if (compareAndSetState(0, 1)) {
> > +                return 1;
> > +            } else {
>
> Re: "(3) Subclass AQS, and provide methods to getValue() / setValue(long).
> We could bitmask the int to store the permit in the upper bits, allowing
> just the use of a single CAS."
>
> I think this is what Jason was talking about in his comment mentioning
> AbstractQueuedLongSynchronizer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/wildfly/wildfly-core/pull/1453/files#r55900166>.
>